flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/logical/utils/LogicalTypeCasts.java

```diff
@@ -0,0 +1,354 @@
+ *
+ * <p>This class aims to be compatible with the SQL standard cast specification.
+ *
+ * <p>Conversions that are defined by the {@link LogicalType} (e.g. interpreting a {@link DateType}
+ * as integer value) are not considered here. They are an internal bridging feature that is not
+ * standard compliant. If at all, {@code CONVERT} methods should make such conversions available.
+ */
+@Internal
+public final class LogicalTypeCasts {
+
+	private static final Map<LogicalTypeRoot, Set<LogicalTypeRoot>> implicitCastingRules;
+
+	private static final Map<LogicalTypeRoot, Set<LogicalTypeRoot>> explicitCastingRules;
+
+	static {
+		implicitCastingRules = new HashMap<>();
+		explicitCastingRules = new HashMap<>();
+
+		// identity casts
+
+		for (LogicalTypeRoot typeRoot : allTypes()) {
+			castTo(typeRoot)
+				.implicitFrom(typeRoot)
+				.build();
+		}
+
+		// cast specification
+
+		castTo(CHAR)
+			.implicitFrom(CHAR)
+			.explicitFromFamily(PREDEFINED)
+			.explicitNotFromFamily(BINARY_STRING)
+			.build();
+
+		castTo(VARCHAR)
+			.implicitFromFamily(CHARACTER_STRING)
+			.explicitFromFamily(PREDEFINED)
+			.explicitNotFromFamily(BINARY_STRING)
+			.build();
+
+		castTo(BOOLEAN)
+			.implicitFrom(BOOLEAN)
+			.explicitFromFamily(CHARACTER_STRING)
+			.build();
+
+		castTo(BINARY)
+			.implicitFromFamily(BINARY_STRING)
+			.build();
+
+		castTo(VARBINARY)
+			.implicitFromFamily(BINARY_STRING)
+			.build();
+
+		castTo(DECIMAL)
+			.implicitFromFamily(NUMERIC)
+			.explicitFromFamily(CHARACTER_STRING)
+			.build();
+
+		castTo(TINYINT)
+			.implicitFrom(TINYINT)
+			.explicitFromFamily(NUMERIC, CHARACTER_STRING)
+			.build();
+
+		castTo(SMALLINT)
+			.implicitFrom(TINYINT, SMALLINT)
+			.explicitFromFamily(NUMERIC, CHARACTER_STRING)
+			.build();
+
+		castTo(INTEGER)
+			.implicitFrom(TINYINT, SMALLINT, INTEGER)
+			.explicitFromFamily(NUMERIC, CHARACTER_STRING)
+			.build();
+
+		castTo(BIGINT)
+			.implicitFrom(TINYINT, SMALLINT, INTEGER, BIGINT)
+			.explicitFromFamily(NUMERIC, CHARACTER_STRING)
+			.build();
+
+		castTo(FLOAT)
+			.implicitFrom(TINYINT, SMALLINT, INTEGER, BIGINT, FLOAT, DECIMAL)
+			.explicitFromFamily(NUMERIC, CHARACTER_STRING)
+			.build();
+
+		castTo(DOUBLE)
+			.implicitFrom(TINYINT, SMALLINT, INTEGER, BIGINT, FLOAT, DOUBLE, DECIMAL)
+			.explicitFromFamily(CHARACTER_STRING)
+			.build();
+
+		castTo(DATE)
+			.implicitFrom(DATE, TIMESTAMP_WITHOUT_TIME_ZONE)
+			.explicitFromFamily(TIMESTAMP, CHARACTER_STRING)
+			.build();
+
+		castTo(TIME_WITHOUT_TIME_ZONE)
+			.implicitFrom(TIME_WITHOUT_TIME_ZONE, TIMESTAMP_WITHOUT_TIME_ZONE)
+			.explicitFromFamily(TIME, TIMESTAMP, CHARACTER_STRING)
+			.build();
+
+		castTo(TIMESTAMP_WITHOUT_TIME_ZONE)
+			.implicitFrom(TIMESTAMP_WITHOUT_TIME_ZONE)
+			.explicitFromFamily(DATETIME, CHARACTER_STRING)
+			.build();
+
+		castTo(TIMESTAMP_WITH_TIME_ZONE)
+			.implicitFrom(TIMESTAMP_WITH_TIME_ZONE)
+			.explicitFromFamily(DATETIME, CHARACTER_STRING)
```

Comment:
Yes, it will be some local type zone defined in the configuration. Oracle has a similar behavior:

`SELECT CAST(TIMESTAMP '2991-12-12 12:12:12' AS TIMESTAMP WITH TIME ZONE) FROM DUAL;`
leads to
`12-DEC-91 12.12.12.000000 PM US/PACIFIC`